File: android/src/main/java/com/mackenrow/nativecard/MackenrowNativeCardView.kt
package com.mackenrow.nativecard

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.os.Build
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.util.TypedValue
import android.view.ActionMode
import android.view.Gravity
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.widget.EditText
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

/**
 * One sensitive input only. React Native owns all layout/chrome/labels/errors.
 * This class is package-private in practice: the JS entry point never exports it.
 */
internal class MackenrowNativeCardView(
  context: Context,
  appContext: AppContext
) : ExpoView(context, appContext), SensitiveFieldHandle {
  private val input = NoExportEditText(context)
  private var sessionId: String? = null
  private var configuredField = SensitiveField.NUMBER
  private var changingText = false
  private var pastedEdit = false
  private var touched = false
  private var revealError = false
  private var autoAdvanced = false
  private var previousIntrinsicStatus = FieldStatus.EMPTY
  private var lastEvent: Map<String, Any>? = null
  private var accessibilityLabelText = ""
  private var accessibilityHintText = ""
  private var enteredAccessibilityValue = ""
  private var invalidAccessibilityValue = ""
  private var placeholderColor = Color.GRAY
  private var fontSizeSp = 16f
  private var fontFamilyName: String? = null
  private var fontWeightName: String? = null
  private var fontStyleName: String? = null
  private var letterSpacingPx: Float? = null

  val onStateChange by EventDispatcher<Map<String, Any>>()

  override val sensitiveField: SensitiveField
    get() = configuredField

  init {
    addView(input, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
    // A plain EditText inflates the host app's `android:editTextBackground`
    // (AppCompat's Material underline, tinted by colorControlNormal/Activated),
    // which the host theme doesn't own or expect. React Native owns all
    // chrome for this field per the class doc above, so strip it — the
    // Android counterpart of iOS's `input.borderStyle = .none`.
    input.background = null
    minimumHeight = dp(48)
    input.minHeight = dp(48)
    input.setSingleLine(true)
    input.setSaveEnabled(false)
    input.isSaveFromParentEnabled = false
    input.setSelectAllOnFocus(false)
    input.customSelectionActionModeCallback = BlockCopyCutActionMode()
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
      input.customInsertionActionModeCallback = BlockCopyCutActionMode()
    }
    input.setOnFocusChangeListener { _, focused ->
      if (!focused) {
        touched = true
        revealError = true
      }
      emitSanitizedState()
    }
    input.addTextChangedListener(object : TextWatcher {
      override fun beforeTextChanged(value: CharSequence?, start: Int, count: Int, after: Int) = Unit

      override fun onTextChanged(value: CharSequence?, start: Int, before: Int, count: Int) {
        if (!changingText) pastedEdit = count > 1
      }

      override fun afterTextChanged(value: Editable?) {
        if (changingText) return
        normalizeTextPreservingCursor()
        val current = intrinsicStatus()
        if (current != FieldStatus.VALID) autoAdvanced = false
        val shouldAdvance =
          previousIntrinsicStatus != FieldStatus.VALID &&
            current == FieldStatus.VALID &&
            !pastedEdit &&
            !autoAdvanced
        previousIntrinsicStatus = current
        if (shouldAdvance) {
          autoAdvanced = true
          sessionId?.let { CardSessionRegistry.focusNext(it, configuredField) }
        } else if (pastedEdit && current == FieldStatus.VALID) {
          autoAdvanced = true
        }
        pastedEdit = false
        emitSanitizedState()
      }
    })
    configureForField()
  }

  fun setSessionId(value: String) {
    if (sessionId == value) return
    sessionId?.let { CardSessionRegistry.unregister(it, this) }
    sessionId = value
    CardSessionRegistry.register(value, this)
  }

  fun setField(value: String) {
    val next = SensitiveField.fromWireName(value) ?: return
    if (configuredField == next) return
    sessionId?.let { CardSessionRegistry.unregister(it, this) }
    clearSensitiveValue()
    configuredField = next
    configureForField()
    sessionId?.let { CardSessionRegistry.register(it, this) }
    emitSanitizedState(force = true)
  }

  fun setDisabled(value: Boolean) {
    input.isEnabled = !value
  }

  fun setPlaceholder(value: String) {
    input.hint = value
    input.setHintTextColor(placeholderColor)
  }

  fun setAccessibilityLabel(value: String) {
    accessibilityLabelText = value
    updateAccessibilityDescription()
  }

  fun setAccessibilityHint(value: String) {
    accessibilityHintText = value
    updateAccessibilityDescription()
  }

  fun setEnteredAccessibilityValue(value: String) {
    enteredAccessibilityValue = value
    updateAccessibilityDescription()
  }

  fun setInvalidAccessibilityValue(value: String) {
    invalidAccessibilityValue = value
    updateAccessibilityDescription()
  }

  fun setTextColor(value: Int) {
    input.setTextColor(value)
  }

  fun setPlaceholderColor(value: Int) {
    placeholderColor = value
    input.setHintTextColor(value)
  }

  fun setCursorColor(value: Int) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) input.textCursorDrawable?.setTint(value)
  }

  fun setFontSize(value: Float) {
    fontSizeSp = value
    input.setTextSize(TypedValue.COMPLEX_UNIT_SP, value)
    applyLetterSpacing()
  }

  fun setFontFamily(value: String?) {
    fontFamilyName = value
    applyTypeface()
  }

  fun setFontWeight(value: String?) {
    fontWeightName = value
    applyTypeface()
  }

  fun setFontStyle(value: String?) {
    fontStyleName = value
    applyTypeface()
  }

  fun setTextAlign(value: String?) {
    input.gravity = Gravity.CENTER_VERTICAL or when (value) {
      "center" -> Gravity.CENTER_HORIZONTAL
      "right" -> Gravity.END
      "left" -> Gravity.START
      else -> Gravity.START
    }
  }

  fun setLetterSpacing(value: Float?) {
    letterSpacingPx = value
    applyLetterSpacing()
  }

  private fun applyTypeface() {
    val weight = fontWeightName?.toIntOrNull()
    val style = (if (weight != null && weight >= 600 || fontWeightName == "bold") {
      Typeface.BOLD
    } else {
      Typeface.NORMAL
    }) or if (fontStyleName == "italic") Typeface.ITALIC else Typeface.NORMAL
    val family = fontFamilyName
    input.typeface = if (family.isNullOrBlank()) {
      Typeface.create(Typeface.DEFAULT, style)
    } else {
      Typeface.create(family, style)
    }
  }

  private fun applyLetterSpacing() {
    input.letterSpacing = letterSpacingPx?.div(fontSizeSp)?.takeIf { it.isFinite() } ?: 0f
  }

  override fun onDetachedFromWindow() {
    sessionId?.let { CardSessionRegistry.unregister(it, this) }
    clearSensitiveValue()
    super.onDetachedFromWindow()
  }

  override fun sensitiveDigits(): String = digits(input.text?.toString().orEmpty())

  override fun isSensitiveValid(): Boolean = intrinsicStatus() == FieldStatus.VALID

  override fun clearSensitiveValue() {
    changingText = true
    input.text?.clear()
    changingText = false
    touched = false
    revealError = false
    autoAdvanced = false
    previousIntrinsicStatus = FieldStatus.EMPTY
    emitSanitizedState(force = true)
  }

  override fun focusSensitiveField() {
    if (input.isEnabled) input.requestFocus()
  }

  override fun blurSensitiveField() {
    input.clearFocus()
  }

  override fun revealValidationError() {
    touched = true
    revealError = true
    emitSanitizedState(force = true)
  }

  private fun configureForField() {
    input.inputType = InputType.TYPE_CLASS_NUMBER or when (configuredField) {
      SensitiveField.CVC -> InputType.TYPE_NUMBER_VARIATION_PASSWORD
      else -> InputType.TYPE_NUMBER_VARIATION_NORMAL
    }
    input.importantForAutofill = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
      IMPORTANT_FOR_AUTOFILL_AUTO
    } else if (configuredField == SensitiveField.NUMBER) {
      input.setAutofillHints("creditCardNumber")
      IMPORTANT_FOR_AUTOFILL_YES
    } else {
      input.setAutofillHints()
      IMPORTANT_FOR_AUTOFILL_NO_EXCLUDE_DESCENDANTS
    }
  }

  private fun normalizeTextPreservingCursor() {
    val currentText = input.text?.toString().orEmpty()
    val oldSelection = input.selectionStart.coerceAtLeast(0)
    val logicalCursor = currentText.take(oldSelection).count(Char::isDigit)
    val normalized = when (configuredField) {
      SensitiveField.NUMBER -> formatNumber(digits(currentText).take(19))
      SensitiveField.EXPIRY -> formatExpiry(digits(currentText).take(4))
      SensitiveField.CVC -> digits(currentText).take(4)
    }
    if (normalized == currentText) return
    changingText = true
    input.setText(normalized)
    input.setSelection(cursorOffsetForDigit(normalized, logicalCursor))
    changingText = false
  }

  private fun formatNumber(value: String): String {
    if (brand(value) == "amex") {
      return buildString {
        value.forEachIndexed { index, char ->
          if (index == 4 || index == 10) append(' ')
          append(char)
        }
      }
    }
    return value.chunked(4).joinToString(" ")
  }

  private fun formatExpiry(value: String) =
    if (value.length > 2) "${value.take(2)}/${value.drop(2)}" else value

  private fun cursorOffsetForDigit(value: String, logicalDigit: Int): Int {
    if (logicalDigit <= 0) return 0
    var seen = 0
    value.forEachIndexed { index, char ->
      if (char.isDigit()) seen += 1
      if (seen == logicalDigit) return index + 1
    }
    return value.length
  }

  // The three functions below (and CardValidation itself) hold every branch
  // of "is this PAN/expiry/CVC complete, valid, or invalid" — deliberately
  // pure and free of `input`/`sessionId` so they're covered by plain JUnit
  // tests instead of only being exercised by hand on a device.
  private fun intrinsicStatus(): FieldStatus {
    val value = sensitiveDigits()
    return when (configuredField) {
      SensitiveField.NUMBER -> CardValidation.numberStatus(value)
      SensitiveField.EXPIRY -> CardValidation.expiryStatus(value)
      SensitiveField.CVC -> CardValidation.cvcStatus(value, CardSessionRegistry.brand(sessionId))
    }
  }

  private fun presentedStatus(): FieldStatus {
    val intrinsic = intrinsicStatus()
    return if (intrinsic == FieldStatus.INVALID && input.hasFocus() && !revealError) {
      FieldStatus.INCOMPLETE
    } else {
      intrinsic
    }
  }

  private fun brand(value: String) = CardValidation.normalizedBrand(value)

  private fun emitSanitizedState(force: Boolean = false) {
    val payload = mapOf(
      "field" to configuredField.wireName,
      "status" to presentedStatus().wireName,
      "brand" to if (configuredField == SensitiveField.NUMBER) brand(sensitiveDigits()) else "unknown",
      "focused" to input.hasFocus(),
      "touched" to touched
    )
    if (force || payload != lastEvent) {
      lastEvent = payload
      updateAccessibilityDescription()
      onStateChange(payload)
    }
  }

  private fun updateAccessibilityDescription() {
    val parts = mutableListOf(accessibilityLabelText)
    if (sensitiveDigits().isNotEmpty()) parts += enteredAccessibilityValue
    if (presentedStatus() == FieldStatus.INVALID) parts += invalidAccessibilityValue
    parts += accessibilityHintText
    input.contentDescription = parts.filter(String::isNotBlank).joinToString(". ")
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) input.tooltipText = accessibilityHintText
  }

  private fun digits(value: String) = value.filter(Char::isDigit)

  private fun dp(value: Int) = (value * resources.displayMetrics.density).toInt()

  private class BlockCopyCutActionMode : ActionMode.Callback {
    override fun onCreateActionMode(mode: ActionMode?, menu: Menu?): Boolean {
      menu?.removeItem(android.R.id.copy)
      menu?.removeItem(android.R.id.cut)
      return true
    }

    override fun onPrepareActionMode(mode: ActionMode?, menu: Menu?): Boolean {
      menu?.removeItem(android.R.id.copy)
      menu?.removeItem(android.R.id.cut)
      return true
    }

    override fun onActionItemClicked(mode: ActionMode?, item: MenuItem?) =
      item?.itemId == android.R.id.copy || item?.itemId == android.R.id.cut

    override fun onDestroyActionMode(mode: ActionMode?) = Unit
  }

  private class NoExportEditText(context: Context) : EditText(context) {
    override fun onTextContextMenuItem(id: Int): Boolean {
      if (id == android.R.id.copy || id == android.R.id.cut) return false
      return super.onTextContextMenuItem(id)
    }

    override fun onKeyShortcut(keyCode: Int, event: KeyEvent): Boolean {
      if (event.isCtrlPressed && (keyCode == KeyEvent.KEYCODE_C || keyCode == KeyEvent.KEYCODE_X)) {
        return true
      }
      return super.onKeyShortcut(keyCode, event)
    }

    override fun onSaveInstanceState(): android.os.Parcelable? = null
  }
}

File: ios/MackenrowNativeCardView.swift
import ExpoModulesCore
import StripePayments
import UIKit

private final class NoExportTextField: UITextField {
  override func canPerformAction(_ action: Selector, withSender sender: Any?) -> Bool {
    if action == #selector(copy(_:)) || action == #selector(cut(_:)) {
      return false
    }
    return super.canPerformAction(action, withSender: sender)
  }

  override func encodeRestorableState(with coder: NSCoder) {
    // Sensitive text is intentionally excluded from UIKit state restoration.
  }
}

/**
 * A single sensitive input. React Native owns layout, chrome, labels and errors.
 * The package's JS entry point never exports this native view.
 */
public final class MackenrowNativeCardView: ExpoView, UITextFieldDelegate, SensitiveFieldHandle {
  private let input = NoExportTextField()
  private var sessionId: String?
  private var configuredField = SensitiveField.number
  private var touched = false
  private var revealError = false
  private var autoAdvanced = false
  private var previousIntrinsicStatus = FieldStatus.empty
  private var lastEvent: [String: AnyHashable]?
  private var fontSize: CGFloat = 17
  private var fontFamily: String?
  private var fontWeight: String?
  private var fontStyle: String?
  private var textAlignment: String?
  private var letterSpacing: CGFloat?
  private var placeholderColor: UIColor = .placeholderText
  private var placeholderText = ""
  private var enteredAccessibilityValue = ""
  private var invalidAccessibilityValue = ""

  public let onStateChange = EventDispatcher()

  internal var sensitiveField: SensitiveField { configuredField }

  public required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = false
    restorationIdentifier = nil
    input.translatesAutoresizingMaskIntoConstraints = false
    input.delegate = self
    input.borderStyle = .none
    input.adjustsFontForContentSizeCategory = true
    input.clearButtonMode = .never
    input.autocorrectionType = .no
    input.spellCheckingType = .no
    input.smartDashesType = .no
    input.smartQuotesType = .no
    input.smartInsertDeleteType = .no
    addSubview(input)
    NSLayoutConstraint.activate([
      input.leadingAnchor.constraint(equalTo: leadingAnchor),
      input.trailingAnchor.constraint(equalTo: trailingAnchor),
      input.topAnchor.constraint(equalTo: topAnchor),
      input.bottomAnchor.constraint(equalTo: bottomAnchor),
      input.heightAnchor.constraint(greaterThanOrEqualToConstant: 44)
    ])
    applyFieldConfiguration()
    applyFont()
  }

  public func setSessionId(_ value: String) {
    guard sessionId != value else { return }
    if let sessionId { CardSessionRegistry.shared.unregister(self, sessionId: sessionId) }
    sessionId = value
    CardSessionRegistry.shared.register(self, sessionId: value)
  }

  public func setField(_ value: String) {
    guard let next = SensitiveField(rawValue: value), configuredField != next else { return }
    if let sessionId { CardSessionRegistry.shared.unregister(self, sessionId: sessionId) }
    clearSensitiveValue()
    configuredField = next
    applyFieldConfiguration()
    if let sessionId { CardSessionRegistry.shared.register(self, sessionId: sessionId) }
    emitSanitizedState(force: true)
  }

  public func setDisabled(_ value: Bool) {
    input.isEnabled = !value
  }

  public func setPlaceholder(_ value: String) {
    placeholderText = value
    input.attributedPlaceholder = nil
    input.placeholder = value
    applyPlaceholder()
  }

  public func setAccessibilityLabel(_ value: String) {
    input.accessibilityLabel = value
  }

  public func setAccessibilityHint(_ value: String) {
    input.accessibilityHint = value
  }

  public func setEnteredAccessibilityValue(_ value: String) {
    enteredAccessibilityValue = value
    updateProtectedAccessibilityValue()
  }

  public func setInvalidAccessibilityValue(_ value: String) {
    invalidAccessibilityValue = value
    updateProtectedAccessibilityValue()
  }

  public func setTextColor(_ value: UIColor) {
    input.textColor = value
  }

  public func setPlaceholderColor(_ value: UIColor) {
    placeholderColor = value
    applyPlaceholder()
  }

  public func setCursorColor(_ value: UIColor) {
    input.tintColor = value
  }

  public func setFontSize(_ value: Double) {
    fontSize = CGFloat(value)
    applyFont()
  }

  public func setFontFamily(_ value: String?) {
    fontFamily = value
    applyFont()
  }

  public func setFontWeight(_ value: String?) {
    fontWeight = value
    applyFont()
  }

  public func setFontStyle(_ value: String?) {
    fontStyle = value
    applyFont()
  }

  public func setTextAlign(_ value: String?) {
    textAlignment = value
    switch value {
    case "center":
      input.textAlignment = .center
    case "right":
      input.textAlignment = .right
    case "left":
      input.textAlignment = .left
    default:
      input.textAlignment = .natural
    }
  }

  public func setLetterSpacing(_ value: Double?) {
    letterSpacing = value.map { CGFloat($0) }
    input.defaultTextAttributes[.kern] = letterSpacing ?? 0
    input.typingAttributes = input.defaultTextAttributes
    applyPlaceholder()
  }

  public override func didMoveToWindow() {
    super.didMoveToWindow()
    if window == nil {
      if let sessionId { CardSessionRegistry.shared.unregister(self, sessionId: sessionId) }
      clearSensitiveValue()
    } else if let sessionId {
      CardSessionRegistry.shared.register(self, sessionId: sessionId)
    }
  }

  internal func sensitiveDigits() -> String {
    STPCardValidator.sanitizedNumericString(for: input.text ?? "")
  }

  internal func isSensitiveValid() -> Bool {
    intrinsicStatus() == .valid
  }

  internal func clearSensitiveValue() {
    input.text = nil
    touched = false
    revealError = false
    autoAdvanced = false
    previousIntrinsicStatus = .empty
    updateProtectedAccessibilityValue()
    emitSanitizedState(force: true)
  }

  internal func focusSensitiveField() {
    if input.isEnabled { input.becomeFirstResponder() }
  }

  internal func blurSensitiveField() {
    input.resignFirstResponder()
  }

  internal func revealValidationError() {
    touched = true
    revealError = true
    emitSanitizedState(force: true)
  }

  public func textFieldDidBeginEditing(_ textField: UITextField) {
    emitSanitizedState()
  }

  public func textFieldDidEndEditing(_ textField: UITextField) {
    touched = true
    revealError = true
    emitSanitizedState()
  }

  public func textField(
    _ textField: UITextField,
    shouldChangeCharactersIn range: NSRange,
    replacementString string: String
  ) -> Bool {
    guard
      let current = textField.text,
      let swiftRange = Range(range, in: current)
    else { return false }

    let prospective = current.replacingCharacters(in: swiftRange, with: string)
    let digitsBeforeRange = current[..<swiftRange.lowerBound].filter(\.isNumber).count
    let insertedDigits = string.filter(\.isNumber).count
    let logicalCursor = digitsBeforeRange + insertedDigits
    let formatted = formattedValue(prospective)
    textField.text = formatted
    if let position = textField.position(
      from: textField.beginningOfDocument,
      offset: cursorOffset(in: formatted, logicalDigit: logicalCursor)
    ) {
      textField.selectedTextRange = textField.textRange(from: position, to: position)
    }

    let currentStatus = intrinsicStatus()
    let isPaste = string.count > 1
    if currentStatus != .valid { autoAdvanced = false }
    let shouldAdvance =
      previousIntrinsicStatus != .valid &&
      currentStatus == .valid &&
      !isPaste &&
      !autoAdvanced
    previousIntrinsicStatus = currentStatus
    if shouldAdvance {
      autoAdvanced = true
      if let sessionId { CardSessionRegistry.shared.focusNext(sessionId, after: configuredField) }
    } else if isPaste && currentStatus == .valid {
      autoAdvanced = true
    }
    updateProtectedAccessibilityValue()
    emitSanitizedState()
    return false
  }

  private func applyFieldConfiguration() {
    input.keyboardType = .numberPad
    input.isSecureTextEntry = configuredField == .cvc
    input.textContentType = configuredField == .number ? .creditCardNumber : .none
  }

  private func applyFont() {
    var base = fontFamily
      .flatMap { UIFont(name: $0, size: fontSize) }
      ?? UIFont.systemFont(ofSize: fontSize, weight: resolvedFontWeight())
    var traits = base.fontDescriptor.symbolicTraits
    if fontStyle == "italic" { traits.insert(.traitItalic) }
    if isBold { traits.insert(.traitBold) }
    if let descriptor = base.fontDescriptor.withSymbolicTraits(traits) {
      base = UIFont(descriptor: descriptor, size: fontSize)
    }
    input.font = UIFontMetrics(forTextStyle: .body).scaledFont(for: base)
    input.adjustsFontForContentSizeCategory = true
    input.defaultTextAttributes[.kern] = letterSpacing ?? 0
    input.typingAttributes = input.defaultTextAttributes
    applyPlaceholder()
  }

  private var isBold: Bool {
    guard let fontWeight else { return false }
    return fontWeight == "bold" || (Int(fontWeight) ?? 400) >= 600
  }

  private func resolvedFontWeight() -> UIFont.Weight {
    switch fontWeight {
    case "100": return .ultraLight
    case "200": return .thin
    case "300": return .light
    case "500": return .medium
    case "600": return .semibold
    case "700", "bold": return .bold
    case "800": return .heavy
    case "900": return .black
    default: return .regular
    }
  }

  private func applyPlaceholder() {
    let value = placeholderText
    guard !value.isEmpty else {
      input.attributedPlaceholder = nil
      return
    }
    var attributes: [NSAttributedString.Key: Any] = [.foregroundColor: placeholderColor]
    attributes[.kern] = letterSpacing ?? 0
    attributes[.font] = input.font ?? UIFont.systemFont(ofSize: fontSize)
    input.placeholder = nil
    input.attributedPlaceholder = NSAttributedString(string: value, attributes: attributes)
  }

  private func formattedValue(_ value: String) -> String {
    let raw = STPCardValidator.sanitizedNumericString(for: value)
    switch configuredField {
    case .number:
      let limited = String(raw.prefix(19))
      if normalizedBrand(limited) == "amex" {
        return limited.enumerated().reduce(into: "") { result, pair in
          if pair.offset == 4 || pair.offset == 10 { result.append(" ") }
          result.append(pair.element)
        }
      }
      return stride(from: 0, to: limited.count, by: 4).map { offset in
        let start = limited.index(limited.startIndex, offsetBy: offset)
        let end = limited.index(start, offsetBy: min(4, limited.distance(from: start, to: limited.endIndex)))
        return String(limited[start..<end])
      }.joined(separator: " ")
    case .expiry:
      let limited = String(raw.prefix(4))
      return limited.count > 2 ? "\(limited.prefix(2))/\(limited.dropFirst(2))" : limited
    case .cvc:
      return String(raw.prefix(4))
    }
  }

  private func cursorOffset(in value: String, logicalDigit: Int) -> Int {
    guard logicalDigit > 0 else { return 0 }
    var seen = 0
    for (offset, character) in value.enumerated() {
      if character.isNumber { seen += 1 }
      if seen == logicalDigit { return offset + 1 }
    }
    return value.count
  }

  // The switch below (and CardValidation itself) holds every branch of "is
  // this PAN/expiry/CVC complete, valid, or invalid" — deliberately pure and
  // free of `input`/`sessionId` so it's covered by plain XCTest instead of
  // only being exercised by hand on a device.
  private func intrinsicStatus() -> FieldStatus {
    let value = sensitiveDigits()
    switch configuredField {
    case .number:
      return CardValidation.numberStatus(value)
    case .expiry:
      return CardValidation.expiryStatus(value)
    case .cvc:
      return CardValidation.cvcStatus(value, brand: CardSessionRegistry.shared.brand(sessionId))
    }
  }

  private func presentedStatus() -> FieldStatus {
    let intrinsic = intrinsicStatus()
    return intrinsic == .invalid && input.isFirstResponder && !revealError ? .incomplete : intrinsic
  }

  private func normalizedBrand(_ value: String) -> String {
    CardValidation.normalizedBrand(value)
  }

  private func updateProtectedAccessibilityValue() {
    var parts: [String] = []
    if !sensitiveDigits().isEmpty { parts.append(enteredAccessibilityValue) }
    if presentedStatus() == .invalid { parts.append(invalidAccessibilityValue) }
    input.accessibilityValue = parts.filter { !$0.isEmpty }.joined(separator: ". ")
  }

  private func emitSanitizedState(force: Bool = false) {
    let payload: [String: AnyHashable] = [
      "field": configuredField.rawValue,
      "status": presentedStatus().rawValue,
      "brand": configuredField == .number ? normalizedBrand(sensitiveDigits()) : "unknown",
      "focused": input.isFirstResponder,
      "touched": touched
    ]
    if force || payload != lastEvent {
      lastEvent = payload
      updateProtectedAccessibilityValue()
      onStateChange(payload)
    }
  }
}
